plugins/modules/s3_object.py

```diff
@@ -917,6 +954,243 @@ def ensure_tags(client, module, bucket, obj):
+                download_s3str(module, connection, s3_vars.bucket, s3_vars.object, version=s3_vars.version)
+        elif s3_vars.version:
+            module.fail_json(msg="Key %s with version id %s does not exist." % (s3_vars.object, s3_vars.version))
+        else:
+            module.fail_json(msg="Key %s does not exist." % s3_vars.object)
+
+
+def s3_object_do_copy(module, connection, s3_vars):
+    # if copying an object in a bucket yet to be created, acls for the bucket and/or the object may be specified
+    # these were separated into the variables bucket_acl and object_acl above
+    d_etag = None
+    if s3_vars.bucketrtn:
+        d_etag = get_etag(connection, s3_vars.bucket, s3_vars.object)
+    else:
+        # If the bucket doesn't exist we should create it.
+        # only use valid bucket acls for create_bucket function
+        s3_vars.permission = s3_vars.bucket_acl
+        create_bucket(module, connection, s3_vars.bucket, s3_vars.location)
+    if not s3_vars.acl_disabled:
+        # only use valid object acls for the copy operation
+        s3_vars.permission = s3_vars.object_acl
+    copy_object_to_bucket(
+        module, connection, s3_vars.bucket,
+        s3_vars.object, s3_vars.encrypt, s3_vars.metadata, s3_vars.validate, d_etag
+    )
+
+class VariableDict(dict):                    
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.__dict__ = self
+    
+    @staticmethod
+    def populate_facts (module):
+            
+        variable_dict = {}
+        for k, v in module.params.items():
+            variable_dict[k] = v
+        variable_dict['object_canned_acl'] = ["private", "public-read", "public-read-write", "aws-exec-read", "authenticated-read", "bucket-owner-read", "bucket-owner-full-control"]
+        variable_dict['bucket_canned_acl'] = ["private", "public-read", "public-read-write", "authenticated-read"]
+
+        if variable_dict['validate_bucket_name']:
+            validate_bucket_name(variable_dict['bucket'])
+
+        if variable_dict.get('overwrite') == 'different' and not HAS_MD5:
+            module.fail_json(msg='overwrite=different is unavailable: ETag calculation requires MD5 support')
+
+        if variable_dict.get('overwrite') not in ['always', 'never', 'different', 'latest']:
+            if module.boolean(variable_dict['overwrite']):
+                variable_dict["overwrite"] = 'always'
+            else:
+                variable_dict["overwrite"] = 'never'
+
+        region, _ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+        # Boto uses symbolic names for locations but region strings will
+        # actually work fine for everything except us-east-1 (US Standard)
+        variable_dict['location'] = region or 'us-east-1'
+
+        if variable_dict.get('object'):
+            # If there is a top level object, do nothing - if the object starts with /
```

Comment:
We don't need to maintain compatibility with Ansible 2.3 and older anymore, a refactor is a good time to remove this IMO.